Multiple offering events

dixita9 opened the pull request:
The multiple offering event feature is mostly done. 

Fixes: #1156

vungc committed:
made modificaitons to the event.py and events.py in models and logic to apply for the new ID added

vungc committed:
made modificaitons to the event.py and events.py in models and logic to apply for the new ID added

MImran2002 committed:
html and js push

MImran2002 committed:
html and js push

MImran2002 committed:
html and js push

MImran2002 committed:
modal toggle button created

MImran2002 committed:
modal toggle button created

MImran2002 committed:
modal toggle button created

MImran2002 committed:
adding modal into the document

MImran2002 committed:
adding modal into the document

MImran2002 committed:
adding modal into the document

vungc committed:
added new tests for isCustom and added isCustom in the lists

vungc committed:
added new tests for isCustom and added isCustom in the lists

vungc committed:
updated functions to resolve isCustom not existing - not solved

vungc committed:
updated functions to resolve isCustom not existing - not solved

MImran2002 committed:
committing again

MImran2002 committed:
committing again

MImran2002 committed:
committing again

vungc committed:
isCustom is passing in the tests, added isCustom in the dictionary in routes, and also added customEventId inside the event table from mysql

vungc committed:
isCustom is passing in the tests, added isCustom in the dictionary in routes, and also added customEventId inside the event table from mysql

vungc committed:
tried to replicate the test for recurring events under test_deleteEvent, but it is not working properly yet for custom events because of saveEventToDb not working propoerly ofr custom events

vungc committed:
tried to replicate the test for recurring events under test_deleteEvent, but it is not working properly yet for custom events because of saveEventToDb not working propoerly ofr custom events

MImran2002 committed:
have done most of the modal ui

MImran2002 committed:
have done most of the modal ui

MImran2002 committed:
have done most of the modal ui

MImran2002 committed:
Merge branch 'imran-annaTraining' of https://github.com/BCStudentSoftwareDevTeam/celts into imran-annaTraining

MImran2002 committed:
morning push

MImran2002 committed:
morning push

MImran2002 committed:
morning push

MImran2002 committed:
commented off all the function and chnages

MImran2002 committed:
commented off all the function and chnages

MImran2002 committed:
commented off all the function and chnages

vungc committed:
copied functions from recurring events to apply to custom events

vungc committed:
copied functions from recurring events to apply to custom events

vungc committed:
copied functions from recurring events to apply to custom events

vungc committed:
copied functions from recurring events to apply to custom events

dixita9 committed:
custom event modal

dixita9 committed:
custom event modal

dixita9 committed:
figured out how to retrieve data from modal

dixita9 committed:
figured out how to retrieve data from modal

dixita9 committed:
figured out how to retrieve data from modal

vungc committed:
created a loop to retriee the data from modal and store it in a list

vungc committed:
created a loop to retriee the data from modal and store it in a list

vungc committed:
created a loop to retriee the data from modal and store it in a list

dixita9 committed:
custom toggle fixed

dixita9 committed:
custom toggle fixed

vungc committed:
fixed the toggle buttons to show custom event headings or recurring event headings based on which one is toggled

vungc committed:
fixed the toggle buttons to show custom event headings or recurring event headings based on which one is toggled

vungc committed:
fixed the toggle buttons to show custom event headings or recurring event headings based on which one is toggled

dixita9 committed:
Save button mostly fixed

dixita9 committed:
Save button mostly fixed

vungc committed:
datepicker shows up when calendar icon is clicked for clones, but only for the most recent clone and not yet working properly

vungc committed:
datepicker shows up when calendar icon is clicked for clones, but only for the most recent clone and not yet working properly

dixita9 committed:
fixed the custom table display

dixita9 committed:
fixed the custom table display

dixita9 committed:
Merge branch 'imran-annaTraining' of https://github.com/BCStudentSoftwareDevTeam/celts into imran-annaTraining

vungc committed:
calendars on the modal are working now

vungc committed:
calendars on the modal are working now

vungc committed:
calendars on the modal are working now

dixita9 committed:
custom toggle fixed

dixita9 committed:
custom toggle fixed

dixita9 committed:
custom toggle fixed

vungc committed:
trash icons are working; clicking on it will delete its corresponding row

vungc committed:
trash icons are working; clicking on it will delete its corresponding row

vungc committed:
trash icons are working; clicking on it will delete its corresponding row

dixita9 committed:
added event name in front end

dixita9 committed:
added event name in front end

dixita9 committed:
custom table completed

dixita9 committed:
custom table completed

dixita9 committed:
custom table completed

dixita9 committed:
hidden input created

dixita9 committed:
hidden input created

vungc committed:
added for loop in the routes to create dictioanries for each of the multiple events

vungc committed:
added for loop in the routes to create dictioanries for each of the multiple events

vungc committed:
chagned time displayed to be in 12hr format, cleaned up HTML and JS files

vungc committed:
chagned time displayed to be in 12hr format, cleaned up HTML and JS files

vungc committed:
chagned time displayed to be in 12hr format, cleaned up HTML and JS files

vungc committed:
tried to debug the dicitoanry in the routes, still work in progress

vungc committed:
tried to debug the dicitoanry in the routes, still work in progress

vungc committed:
tried to debug the dicitoanry in the routes, still work in progress

dixita9 committed:
json error fixed

dixita9 committed:
json error fixed

dixita9 committed:
json error fixed

vungc committed:
data is saving to the databse

vungc committed:
data is saving to the databse

dixita9 committed:
admin logs are working correctly now

dixita9 committed:
admin logs are working correctly now

vungc committed:
cleaned the debug statements in routes and modified the adminlog to display and before the last event and date so that it sounds like an actual sentence

vungc committed:
cleaned the debug statements in routes and modified the adminlog to display and before the last event and date so that it sounds like an actual sentence

vungc committed:
flash messages appear after tojson is added back on base.html, save button on modal is now blue, and custom event IDs now exist

vungc committed:
flash messages appear after tojson is added back on base.html, save button on modal is now blue, and custom event IDs now exist

vungc committed:
flash messages appear after tojson is added back on base.html, save button on modal is now blue, and custom event IDs now exist

vungc committed:
modified customDict in routes to only update the different data

vungc committed:
modified customDict in routes to only update the different data

dixita9 committed:
changed the name of the custom events to multiple offering events

dixita9 committed:
changed the name of the custom events to multiple offering events

dixita9 committed:
changed the name of the custom events to multiple offering events

dixita9 committed:
added multiple offerings to some tests

dixita9 committed:
added multiple offerings to some tests

vungc committed:
added a test for saving multiple offering events

vungc committed:
added a test for saving multiple offering events

dixita9 committed:
tests for multiple offerings completed

dixita9 committed:
tests for multiple offerings completed

dixita9 committed:
pulled from development

JohnCox2211 committed:
modified trash can button to be similar to the
rest of the code base

JohnCox2211 committed:
modified trash can button to be similar to the
rest of the code base

JohnCox2211 committed:
Formatted the add multiple event button to
indicate its purpose better.

JohnCox2211 committed:
Added css file for createEvent for button styling

JohnCox2211 committed:
made logging functions consistent with those in
development

JohnCox2211 committed:
resovled merge issues

JohnCox2211 committed:
resolved merge issues

JohnCox2211 committed:
no more errorr

JohnCox2211 committed:
Added date type to the date entry for default
event page. The date does not appear autofilled
in the edit page

JohnCox2211 committed:
end of day commit, working on default date picker
tomorrow

dixita9 committed:
merged changes

JohnCox2211 committed:
TRYING TO PULL

dixita9 committed:
Merge branch 'development' of https://github.com/BCStudentSoftwareDevTeam/celts into MultipleOfferingEv

dixita9 committed:
cleaned create event page

dixita9 committed:
cleaned create event page

JohnCox2211 committed:
removed dupe divs

JohnCox2211 committed:
removed dupe divs

JohnCox2211 committed:
added required tags in modal

JohnCox2211 committed:
Added function to make modal fields required. Need
to fix the notifier to appear on the modal.

dixita9 committed:
merged

JohnCox2211 committed:
Added a lil text notifier in the modal to tell the
user they have bad input, needs to be beautified

dixita9 committed:
Merge branch 'multipleOfferingJohn' of https://github.com/BCStudentSoftwareDevTeam/celts into MultipleOfferingEvents

dixita9 committed:
designed the message flash for modal

dixita9 committed:
added comments for tomorrow's work

JohnCox2211 committed:
Corrected spelling and centered text for the
in-modal notifier

JohnCox2211 committed:
main page name transfers to modal, working on
recurring and multiple event toggles

JohnCox2211 committed:
User can no longer select recurring event and
multi event. When recurring is toggled on, you
can not toggle multi, but with multi on, you can
toggle recurring on. Ideally the same thing would
happen in either scenario, but it works right now
and the user will get a notifier to tell them that
both can not be toggled.

dixita9 committed:
fixed required inputs in modal: event name

JohnCox2211 committed:
All fields now required in the event modal

JohnCox2211 committed:
fixed modal notifier formatting

JohnCox2211 committed:
add animation to the modal notifier

JohnCox2211 committed:
removed unused function

dixita9 committed:
date formating in process

dixita9 committed:
date formating in process

dixita9 committed:
date formating in process

dixita9 committed:
Merge branch 'multipleOfferingJohn' of https://github.com/BCStudentSoftwareDevTeam/celts into MultipleOfferingEvents

dixita9 committed:
date format for modal table fixed

dixita9 committed:
date format for modal table fixed

dixita9 committed:
date format for modal table fixed

JohnCox2211 committed:
commiting to pull changes

JohnCox2211 committed:
added function back that git deemed useless for
some reason

JohnCox2211 committed:
modified toggle switch logic, still not working
as wanted

dixita9 committed:
muliple offering and recurring toggle cannot be toggled at the same time!

dixita9 committed:
muliple offering and recurring toggle cannot be toggled at the same time!

dixita9 committed:
toggle fixed and branches merged

dixita9 committed:
removed unncessary lines of code from toggle code

dixita9 committed:
removed unncessary lines of code from toggle code

JohnCox2211 committed:
fixed the code that git broke again

dixita9 committed:
date picker fixed

JohnCox2211 committed:
removed dupe function that git decided it wanted
to dupe, reformatted HTML display date

dixita9 committed:
Merge branch 'multipleOfferingJohn' of https://github.com/BCStudentSoftwareDevTeam/celts into MultipleOfferingEvents

dixita9 committed:
fixed dte picker

dixita9 committed:
fixed duplicate code

dixita9 committed:
fixed merge issues

JohnCox2211 committed:
removed more dupes

dixita9 committed:
fixed end date

JohnCox2211 committed:
corrected HTML date formatting for editing event

JohnCox2211 committed:
removed unused code and updated comments in
HTML

JohnCox2211 committed:
No more error with saving events due to the date
previously not being required after toggling
multiple event modal. Start date picker now
reappears when the modal is canceled or X'd out

dixita9 committed:
Merge branch 'multipleOfferingJohn' of https://github.com/BCStudentSoftwareDevTeam/celts into MultipleOfferingEvents

dixita9 committed:
edit button created

BrianRamsay commented on app/controllers/admin/routes.py:
```diff
@@ -98,13 +104,29 @@ def createEvent(templateid, programid):
-            savedEvents, validationErrorMessage = attemptSaveEvent(eventData, getFilesFromRequest(request))
+        if eventData.get('isMultipleOffering'):
+            multipleOfferingId = calculateNewMultipleOfferingId()
+            for event in ast.literal_eval(eventData.get('multipleOfferingData')):
```
We should use json encoding and decoding vs python evals

JohnCox2211 replied:
replaced ast evals with json in both spots and removed the ast import

BrianRamsay commented on app/controllers/admin/routes.py:
```diff
@@ -5,6 +5,10 @@
+import ast
+from dateutil.parser import parse
+import ast
+from dateutil.parser import parse
```
Duplicate imports.

BrianRamsay replied:
Also it doesn't look like we use `parse`

JohnCox2211 replied:
removed these imports all together.

BrianRamsay commented on app/controllers/admin/routes.py:
```diff
@@ -113,14 +135,33 @@ def createEvent(templateid, programid):
+                elif len(savedEventsList) >= 1 and eventData.get('isMultipleOffering'):
+                    modifiedSavedEvents = [item for sublist in savedEventsList for item in sublist]
+                    
+                    event_dates = [datetime.strptime(event_data['eventDate'], '%Y-%m-%d').date().strftime('%m/%d/%Y') for event_data in ast.literal_eval(eventData.get('multipleOfferingData'))]
```
Don't use user input this far down in the function. You've already gotten the data structure once.

Also, in this case, we should use the dates from the saved events, not from the event data that was passed in.

JohnCox2211 replied:
now using saved event list to get the event date

JohnCox2211 committed:
added background colors to the event tables and
multi event modal.

JohnCox2211 committed:
added background colors to the event tables and
multi event modal.

JohnCox2211 committed:
merged branch changes

JohnCox2211 committed:
merged branch changes

JohnCox2211 committed:
merged branch changes

JohnCox2211 committed:
merged branch changes

BrianRamsay commented on app/controllers/admin/routes.py:
```diff
@@ -113,14 +135,33 @@ def createEvent(templateid, programid):
+
+                    if len(modifiedSavedEvents) > 1:
+                        #get last event name and stick at the end before 'and' so that it reads like a sentence in admin log
+                        last_event = f"<a href=\"{url_for('admin.eventDisplay', eventId=modifiedSavedEvents[-1].id)}\">{modifiedSavedEvents[-1].name}</a>"
```
We should rework this so we don't have to duplicate the link html

JohnCox2211 replied:
now only using one link, building the list without last_event variable and instead just parsing that and adding it to the end

BrianRamsay commented on app/logic/events.py:
```diff
@@ -67,35 +67,39 @@ def deleteEvent(eventId):
-                allRecurringEvents = list(Event.select().where(Event.recurringId == recurringId))
+            if event.recurringId or event.multipleOfferingId:
+                recurringOrMultipleOfferingId = event.recurringId or event.multipleOfferingId
+                allRecurringEvents = list(Event.select().where(Event.recurringId == recurringOrMultipleOfferingId or Event.multipleOfferingId == recurringOrMultipleOfferingId))
```
This will be a problem if those ids overlap, because you would select unrelated events.

JohnCox2211 replied:
separated recurringOrMultiple valiable, ask for further elaboration

JohnCox2211 committed:
made the padding spacing in the modal better

BrianRamsay commented on app/logic/events.py:
```diff
@@ -136,17 +141,26 @@ def saveEventToDb(newEventData, renewedEvent = False):
                                 'date':newEventData['startDate'],
                                 "week":1})
         if renewedEvent:
             recurringSeriesId = newEventData.get('recurringId')
```
We don't handle the multiple offering case here.

BrianRamsay replied:
Also, that `append` happens in both cases, so we should make the conditional for `isMultipleOffering` only be around the code that depends on the condition.

BrianRamsay commented on app/static/js/createEvents.js:
```diff
@@ -49,47 +51,96 @@ function calculateRecurringEventFrequency(){
-    dateFormat: 'mm/dd/yy',
-    minDate: new Date() 
-  });
+  document.getElementById('submitParticipant').addEventListener('click', function() {   
```
Since we use jquery event listener syntax elsewhere, let's do it here, too.

JohnCox2211 replied:
Changed this to jQuery listener as suggested

BrianRamsay commented on app/static/js/createEvents.js:
```diff
@@ -49,47 +51,96 @@ function calculateRecurringEventFrequency(){
+        $('.invalidFeedback').css('display', 'none');
+        $('#textNotifierPadding').removeClass('pt-5')
+      });
+      isEmpty = false;
```
It would be useful to highlight which input is empty. Making a thicker red border should be fine. You can save the input into isEmpty instead of true, so you have it later.

dixita9 committed:
fixed the functionality of edit button

dixita9 committed:
fixed the functionality of edit button

dixita9 committed:
fixed the functionality of edit button

dixita9 committed:
fixed the functionality of edit button

dixita9 committed:
fixed the functionality of edit button

BrianRamsay commented on app/static/js/createEvents.js:
```diff
@@ -49,47 +51,96 @@ function calculateRecurringEventFrequency(){
-  $(".startDatePicker, .endDatePicker").change(function () {
-    updateDate(this);
+    else {
+      storingMultipleOfferingEventAttributes();
```
To match our other functions, this should probably be named `storeMultipleOfferingEventAttributes`

JohnCox2211 replied:
made the suggested change

BrianRamsay commented on app/static/js/createEvents.js:
```diff
@@ -49,47 +51,96 @@ function calculateRecurringEventFrequency(){
+    let entries = [];
+    $(".extraSlots").children().each(function(index, element) {
+        let rowData = $.map($(element).find("input"), (el) => $(el).val());
+        console.log("Data in row " + (index + 1) + ": " + rowData);
```
We don't need debugging output anymore

JohnCox2211 replied:
removed console log

BrianRamsay commented on app/static/js/createEvents.js:
```diff
@@ -49,47 +51,96 @@ function calculateRecurringEventFrequency(){
+    });
+
+    let entriesJson = JSON.stringify(entries);
+    document.getElementById("multipleOfferingDataId").value = entriesJson
```
We use jquery everywhere else, let's use it here.

JohnCox2211 replied:
replaced default JS with jQuery here and in other needed locations.

BrianRamsay commented on app/static/js/createEvents.js:
```diff
@@ -103,20 +154,78 @@ $(document).ready(function () {
+      document.getElementById('eventName').value = document.getElementById('inputEventName').value; //keeps main page event name for multiple event modal
+    }
+    var recurringStatus = $("input[id='checkIsRecurring']:checked").val(); //this line function is to retrive ON when its toggle for recurring event on createEvent.html line 158
+    var multipleOfferingStatus = $("input[id='checkIsMultipleOffering']:checked").val();// this line function is to retrive ON when toggle for multiple offering event button createEvent.html line 160
```
Let's put all of these end-of-line comments before the lines instead of at the end. It's easier to follow.

BrianRamsay replied:
Also, these selectors are complicated. we don't need to use an attribute selector if we know an id, we can just use the id. That is, instead of `$("input[id='checkIsRecurring']:checked")`, use `$("#checkIsRecurring:checked")`.

Or maybe use booleans everywhere instead of 'on' or '', with `$("#checkIsRecurring").is(":checked")`.

JohnCox2211 replied:
replaced current way of checking if these are toggled to now be booleans as suggested.

JohnCox2211 committed:
fixed multi event table date formatting causing
the dates to appear to be a day before the event
was set to be held

BrianRamsay commented on app/static/js/createEvents.js:
```diff
@@ -103,20 +154,78 @@ $(document).ready(function () {
+    }
+    var recurringStatus = $("input[id='checkIsRecurring']:checked").val(); //this line function is to retrive ON when its toggle for recurring event on createEvent.html line 158
+    var multipleOfferingStatus = $("input[id='checkIsMultipleOffering']:checked").val();// this line function is to retrive ON when toggle for multiple offering event button createEvent.html line 160
+    modalOpenedByEditButton = ($(this).attr('id') === 'edit_modal');
```
`this.id` should work here.

BrianRamsay commented on app/static/js/createEvents.js:
```diff
@@ -103,20 +154,78 @@ $(document).ready(function () {
+    $('#modalMultipleOffering').modal('hide');
+    $('.extraSlots').children().not(':first').remove();
+  }
+  });
```
Same deal with the end-of-line comments in the above chunk of code.

JohnCox2211 replied:
corrected commenting

BrianRamsay commented on app/static/js/createEvents.js:
```diff
@@ -103,20 +154,78 @@ $(document).ready(function () {
+      $('#modalMultipleOffering').modal('show');
+      $('#nonMultipleOfferingTime, #nonMultipleOfferingDate').addClass('d-none'); // this line hides the non multiple offering time and dates and replace them with recurring table div for multiple offering events to show
+    }
+    else{
```
I'd add a comment here that clarifies that this is the recurring case

BrianRamsay commented on app/static/js/createEvents.js:
```diff
@@ -103,20 +154,78 @@ $(document).ready(function () {
+    $("#multipleOfferingTableDiv").addClass('d-none');
+    $('#modalMultipleOffering').modal('hide');
+    $('.extraSlots').children().not(':first').remove();
+  }
```
Fix the indentation please.

JohnCox2211 replied:
reformatted

BrianRamsay commented on app/static/js/createEvents.js:
```diff
@@ -103,20 +154,78 @@ $(document).ready(function () {
+  });
+  
+  /*cloning the div with ID multipleOfferingEvent and cloning, changing the ID of each clone going up by 1. This also changes the ID of the deleteMultipleOfferingEvent so that when the trash icon is clicked, 
+  that specific row will be deleted*/
```
Let's format this comment better and line it up.

```
/* This is a multi-line comment
   that is indented and not too long */
```

JohnCox2211 replied:
reformatted

BrianRamsay commented on app/templates/base.html:
```diff
@@ -48,6 +48,7 @@
           <div class="flash_container" id="flash_container">
             {% with messages = get_flashed_messages(with_categories=true) %}
               <script> var flashMessages = {{messages | tojson}} </script>
+              {# <script> var flashMessages = {{messages | tojson}} </script> #}  {#FIXME: SHOULD BE THIS, WHY??#}
```
Not sure what this comment was doing. I'm guessing you figured this out?

JohnCox2211 replied:
figured it out, removed comment line that was unused

JohnCox2211 committed:
removed dupe code

BrianRamsay reviewed (changes requested):
I haven't finished, but here is some stuff to work on.

For most of the code, let's move your comments at the end of a line to the line before the code in question. It's easier to read in the flow of the code, and doesn't cause a line-wrap

JohnCox2211 committed:
removed dupe code again

JohnCox2211 committed:
removed dupe code again

JohnCox2211 committed:
changed comment placing, changed recurring and
multi event toggles to give booleans

removed HTML base dupe line

removed dupe imports in routes

JohnCox2211 committed:
changed comment placing, changed recurring and
multi event toggles to give booleans

removed HTML base dupe line

removed dupe imports in routes

JohnCox2211 committed:
Updated specified default JS code to be JQuery.

JohnCox2211 committed:
split a variable that would hold recurring event
id or multiple event id. Replaced ast python
literals with json as requested

dixita9 committed:
red border added to the modl inputs that are empty, edit modal button is not working correctly

dixita9 committed:
red border added to the modl inputs that are empty, edit modal button is not working correctly

dixita9 committed:
red border added to the modl inputs that are empty, edit modal button is not working correctly

JohnCox2211 committed:
merged working branches

dixita9 committed:
edit modal issue fixed

JohnCox2211 committed:
removed unused function

JohnCox2211 committed:
added back parenthesis that git removed

JohnCox2211 committed:
corrected variable names

JohnCox2211 committed:
now use saved event dates instead of getting new
 input

JohnCox2211 committed:
updated formatting for date correction

Commit:
Demo commit

dixita9 committed:
checked for end time after start time

dixita9 committed:
Merge branch 'multipleOfferingJohn' of https://github.com/BCStudentSoftwareDevTeam/celts into MultipleOfferingEvents

dixita9 committed:
working fine

dixita9 committed:
added red border to start time and end time

dixita9 committed:
Merge branch 'development' of https://github.com/BCStudentSoftwareDevTeam/celts into MultipleOfferingEvents

dixita9 committed:
Merged with Development

dixita9 committed:
datepicker fixed

dixita9 committed:
fixed footer

dixita9 committed:
fixed footer

Commit:
Solved merge conflicts

Commit:
Save button was not appearing due to an issue in HTML, fixed.

BrianRamsay reviewed (changes requested):
Restore the styles from John's issue in a way that works for both features.

The modal footer has disappeared!

There's one more merge conflict

Commit:
Centred Headings for Optional and Must Select

Commit:
Fixed modal footer issue

Commit:
merge conflict resloved. Footer works as expected. Save and Cancel functionality works like it used to.

BrianRamsay committed:
Merge branch 'development' into MultipleOfferingEvents

Commit:
Removed MultipleOffering id from delete function

Commit:
Merge branch 'MultipleOfferingEvents' of https://github.com/BCStudentSoftwareDevTeam/celts into MultipleOfferingEvents
merging developement into MultipleOfferingEvents

dixita9 committed:
completed
